File: Coding/890/findAndReplacePattern.cpp
class Solution {
public:
    // Time efficiency: O(len(pattern)*len(words)); Space Complexity: O(len(pattern)*len(words))
    // The characters are not important, the important thing is the order. 
    // Therefore, it is easy to turn the characters to numbers and the compare. 
    vector<int> find_pattern(string word) {
        int cur = 0;
        map<char, int> chr_to_int;
        vector<int> ret;
        for(auto c: word){
            if (chr_to_int.find(c) == chr_to_int.end()) {
                chr_to_int[c]=cur;
                cur+=1;
            }
            ret.emplace_back(chr_to_int[c]);
        }
        return ret;
    }
    
    vector<string> findAndReplacePattern(vector<string>& words, string pattern) {
        vector<int> intPattern = find_pattern(pattern);
        vector<string> ret;
        for(auto word: words){
            vector<int> intCurPattern = find_pattern(word);
            if (intCurPattern == intPattern) {
                ret.emplace_back(word);
            }
        }
        return ret;
    }
};

File: Coding/890/findAndReplacePattern.py
class Solution(object):
    # Time efficiency: O(len(words)*len(pattern)), Space efficiency: O(len(pattern)*len(words))
    def findAndReplacePattern(self, words, pattern):
        """
        :type words: List[str]
        :type pattern: str
        :rtype: List[str]
        """
        def ifPattern(word):
            word_to_pattern = {}
            pattern_to_word = {}
            for i in range(len(word)):
                if word[i] not in word_to_pattern and pattern[i] not in pattern_to_word:
                    word_to_pattern[word[i]] = pattern[i]
                    pattern_to_word[pattern[i]] = word[i]
                elif word[i] not in word_to_pattern or pattern[i] not in pattern_to_word:
                    return False
                else:
                    if word_to_pattern[word[i]] != pattern[i] or pattern_to_word[pattern[i]] != word[i]:
                        return False
            return True
        
        ret = []
        for word in words:
            if ifPattern(word):
                ret.append(word)
        return ret
